File: src/test/java/org/nyxlang/interpreter/InterpreterBoolTest.java
package org.nyxlang.interpreter;

import org.junit.jupiter.api.Test;
import org.nyxlang.TestSuit;
import org.nyxlang.interpreter.exception.InterpreterException;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class InterpreterBoolTest extends TestSuit {

    @Test
    void shouldCompareTrueAndFalseKeywords() {
        var result = execute("true == false");
        assertEquals(getFalseList(), result);

        result = execute("true == true");
        assertEquals(getTrueList(), result);

        result = execute("true && true");
        assertEquals(getTrueList(), result);

        result = execute("false && true");
        assertEquals(getFalseList(), result);

        result = execute("false || true");
        assertEquals(getTrueList(), result);

        result = execute("false || (1 == 2)");
        assertEquals(getFalseList(), result);
    }


    @Test
    void shouldComparePositiveNumbers() {
        var result = execute("1<2");
        assertEquals(getTrueList(), result);

        result = execute("1>2");
        assertEquals(getFalseList(), result);
    }

    @Test
    void shouldCompareNumbers() {
        var result = execute("-1<2");
        assertEquals(getTrueList(), result);

        result = execute("1>-2");
        assertEquals(getTrueList(), result);
    }

    @Test
    void shouldNegate() {
        var result = execute("!(-1<2)");
        assertEquals(getFalseList(), result);

        result = execute("!(1>-2)");
        assertEquals(getFalseList(), result);
    }

    @Test
    void shouldCheckEquality() {
        var result = execute("3928==3928");
        assertEquals(getTrueList(), result);

        result = execute("-392==392");
        assertEquals(getFalseList(), result);
    }

    @Test
    void shouldCheckNonEquality() {
        var result = execute("3927!=3928");
        assertEquals(getTrueList(), result);

        result = execute("-392!=392");
        assertEquals(getTrueList(), result);

        result = execute("392!=392");
        assertEquals(getFalseList(), result);

        result = execute("true!=false");
        assertEquals(getTrueList(), result);
    }

    @Test
    void shouldCompareWithLogicalAnd() {
        var result = execute("(1<2)&&(1==1)");
        assertEquals(getTrueList(), result);

        result = execute("(1>2)&&(1==1)");
        assertEquals(getFalseList(), result);

        result = execute("(1>2)&&(1==2)");
        assertEquals(getFalseList(), result);
    }

    @Test
    void shouldCompareWithLogicalOr() {
        var result = execute("(1<2)||(1==1)");
        assertEquals(getTrueList(), result);

        result = execute("(1>2)||(1==1)");
        assertEquals(getTrueList(), result);

        result = execute("(1>2)||(1==2)");
        assertEquals(getFalseList(), result);
    }

    @Test
    void shouldCompareWithGreaterEquals() {
        var result = execute("2>=2");
        assertEquals(getTrueList(), result);

        result = execute("3>=2");
        assertEquals(getTrueList(), result);

        result = execute("1>=2");
        assertEquals(getFalseList(), result);
    }

    @Test
    void shouldCompareWithLessEquals() {
        var result = execute("2<=2");
        assertEquals(getTrueList(), result);

        result = execute("3<=2");
        assertEquals(getFalseList(), result);

        result = execute("1<=2");
        assertEquals(getTrueList(), result);
    }

    @Test
    void shouldErrorForInvalidBoolLogic() {
        assertThrows(InterpreterException.class, () -> execute("true - false"));
        assertThrows(InterpreterException.class, () -> execute("-false"));
    }

    @Test
    void shouldHaveCorrectPrecedenceForEquals() {
        var result = execute("1 < 2 == true");
        assertEquals(getTrueList(), result);

        result = execute("true == 1 < 2");
        assertEquals(getTrueList(), result);
    }
}

File: src/test/kotlin/org/leaflang/interpreter/InterpreterBoolTest.kt
package org.leaflang.interpreter

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.leaflang.TestSuit
import org.leaflang.interpreter.exception.DynamicSemanticException

class InterpreterBoolTest : TestSuit() {
    @Test
    fun shouldCompareTrueAndFalseKeywords() {
        var result = execute("true == false")
        assertFalse(result as Boolean)

        result = execute("true == true")
        assertTrue(result as Boolean)

        result = execute("true && true")
        assertTrue(result as Boolean)

        result = execute("false && true")
        assertFalse(result as Boolean)

        result = execute("false || true")
        assertTrue(result as Boolean)

        result = execute("false || (1 == 2)")
        assertFalse(result as Boolean)
    }

    @Test
    fun shouldComparePositiveNumbers() {
        var result = execute("1<2")
        assertTrue(result as Boolean)

        result = execute("1>2")
        assertFalse(result as Boolean)
    }

    @Test
    fun shouldCompareNumbers() {
        var result = execute("-1<2")
        assertTrue(result as Boolean)

        result = execute("1>-2")
        assertTrue(result as Boolean)
    }

    @Test
    fun shouldNegate() {
        var result = execute("!(-1<2)")
        assertFalse(result as Boolean)

        result = execute("!(1>-2)")
        assertFalse(result as Boolean)
    }

    @Test
    fun shouldCheckEquality() {
        var result = execute("3928==3928")
        assertTrue(result as Boolean)

        result = execute("-392==392")
        assertFalse(result as Boolean)
    }

    @Test
    fun shouldCheckNonEquality() {
        var result = execute("3927!=3928")
        assertTrue(result as Boolean)

        result = execute("-392!=392")
        assertTrue(result as Boolean)

        result = execute("392!=392")
        assertFalse(result as Boolean)

        result = execute("true!=false")
        assertTrue(result as Boolean)
    }

    @Test
    fun shouldCompareWithLogicalAnd() {
        var result = execute("(1<2)&&(1==1)")
        assertTrue(result as Boolean)

        result = execute("(1>2)&&(1==1)")
        assertFalse(result as Boolean)

        result = execute("(1>2)&&(1==2)")
        assertFalse(result as Boolean)
    }

    @Test
    fun shouldCompareWithLogicalOr() {
        var result = execute("(1<2)||(1==1)")
        assertTrue(result as Boolean)

        result = execute("(1>2)||(1==1)")
        assertTrue(result as Boolean)

        result = execute("(1>2)||(1==2)")
        assertFalse(result as Boolean)
    }

    @Test
    fun shouldCompareWithGreaterEquals() {
        var result = execute("2>=2")
        assertTrue(result as Boolean)

        result = execute("3>=2")
        assertTrue(result as Boolean)

        result = execute("1>=2")
        assertFalse(result as Boolean)
    }

    @Test
    fun shouldCompareWithLessEquals() {
        var result = execute("2<=2")
        assertTrue(result as Boolean)

        result = execute("3<=2")
        assertFalse(result as Boolean)

        result = execute("1<=2")
        assertTrue(result as Boolean)
    }

    @Test
    fun shouldErrorForInvalidBoolLogic() {
        assertThrows(DynamicSemanticException::class.java) { execute("true - false") }
        assertThrows(DynamicSemanticException::class.java) { execute("-false") }
    }

    @Test
    fun shouldHaveCorrectPrecedenceForEquals() {
        var result = execute("1 < 2 == true")
        assertTrue(result as Boolean)

        result = execute("true == 1 < 2")
        assertTrue(result as Boolean)
    }
}